### PR TITLE
Markup declared properties

### PR DESCRIPTION
**Pull request by exyi**

### CODE REVIEW

File: src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs [hunk 1]
```diff
@@ -527,6 +527,19 @@ protected override Expression VisitFormattedExpression(FormattedBindingParserNod
...
 
+        protected override Expression VisitArrayInitializer(ArrayInitializerExpression node)
+        {
+            var initializers = node.ElementInitializers.Select(e => Visit(e));
+
+            var firstInitializer = initializers.FirstOrDefault();
+
+            var firstElementType = firstInitializer?.Type ?? typeof(object);
+
+            var arrayElementType = initializers.All(i => i.Type.IsAssignableFrom(firstElementType)) ? firstElementType : typeof(object);
```
Comment (exyi) [inline src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs:538]: I'd prefer to throw an exception if we can't find a good element type. We can always add more handled cases later without breaking anyone. And in unhandled cases it would be better to support C# cast operation instead of doing magic implicit conversions

`[(object) A, (object) B]`


or a cast method like this: `Cast<object>(A)`

File: src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs [hunk 2]
```diff
@@ -527,6 +527,19 @@ protected override Expression VisitFormattedExpression(FormattedBindingParserNod
...
+
+            return Expression.NewArrayInit(arrayElementType, initializers.Select(i => Expression.Convert(i, arrayElementType)));
```
Comment (exyi) [inline src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs:540]: Maybe using `TypeConversions.ImplicitConversion()` instead of Expression.Convert and IsAssignableTo

File: src/Framework/Framework/Compilation/Binding/TypeConversions.cs [hunk 3]
```diff
@@ -158,6 +158,10 @@ public static Expression BoxToObject(Expression src)
             return null;
         }
 
+        //TODO: Refactor ImplicitConversion usages to EnsureImplicitConversion where applicable to take advantage of nullability 
+        public static Expression EnsureImplicitConversion(Expression src, Type destType)
+            => ImplicitConversion(src, destType, true, false)!;
```
Comment (exyi) [inline src/Framework/Framework/Compilation/Binding/TypeConversions.cs:163]: :+1:

File: src/Framework/Framework/Compilation/ControlTree/ControlTreeResolverBase.cs [hunk 4]
```diff
@@ -29,81 +21,59 @@ public abstract class ControlTreeResolverBase : IControlTreeResolver
...
                 );
             }
 
-            ValidateMasterPage(view, masterPage, masterPageDirective?.First());
-
             ResolveRootContent(root, view, viewMetadata);
 
             return view;
-        }
+        }
```
Comment (exyi) [inline src/Framework/Framework/Compilation/ControlTree/ControlTreeResolverBase.cs:76]: :eyes:

File: src/Framework/Framework/Compilation/ControlTree/IAbstractDirectiveAttributeReference.cs [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+using DotVVM.Framework.Compilation.Parser.Binding.Parser;
+
+namespace DotVVM.Framework.Compilation.ControlTree
+{
+    public interface IAbstractDirectiveAttributeReference
+    {
+        public TypeReferenceBindingParserNode TypeSyntax { get; }
+        public IdentifierNameBindingParserNode NameSyntax { get; }
+        public LiteralExpressionBindingParserNode Initializer { get; }
+        public ITypeDescriptor? Type { get; }
```
Comment (exyi) [inline src/Framework/Framework/Compilation/ControlTree/IAbstractDirectiveAttributeReference.cs:12]:
```suggestion
        TypeReferenceBindingParserNode TypeSyntax { get; }
        IdentifierNameBindingParserNode NameSyntax { get; }
        LiteralExpressionBindingParserNode Initializer { get; }
        ITypeDescriptor? Type { get; }
```

Comment (exyi) [reply src/Framework/Framework/Compilation/ControlTree/IAbstractDirectiveAttributeReference.cs:12]: Maybe add a short comment explaining what this node represents. It wasn't clear to me immediately :sweat_smile:

File: src/Framework/Framework/Compilation/ControlTree/Resolved/DirectiveCompilationService.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+            try {
+                var initializerExpression = visitor.Visit(initializer);
+
+                var funcType = typeof(Func<>).MakeGenericType(propertyType);
+
+                var lambda = Expression.Lambda(funcType, Expression.Block(TypeConversion.EnsureImplicitConversion(initializerExpression, propertyType)));
+                var lambdaDelegate = lambda.Compile(true);
+
+                return lambdaDelegate.DynamicInvoke() ?? CreateDefaultValue(propertyType);
```
Comment (exyi) [inline src/Framework/Framework/Compilation/ControlTree/Resolved/DirectiveCompilationService.cs:80]:
```suggestion


                var lambda = Expression.Lambda<Func<object?>>( Expression.Block(Convert(TypeConversion.EnsureImplicitConversion(initializerExpression, propertyType), typeof(object)));
                var lambdaDelegate = lambda.Compile(true);

                return lambdaDelegate.Invoke() ?? CreateDefaultValue(propertyType);
```

File: src/Framework/Framework/Compilation/ControlTree/Resolved/ResolvedPropertyDeclarationDirective.cs [hunk 7]
```diff
@@ -0,0 +1,100 @@
+﻿#nullable enable
```
Comment (exyi) [inline src/Framework/Framework/Compilation/ControlTree/Resolved/ResolvedPropertyDeclarationDirective.cs:1]:
```suggestion
```

File: src/Framework/Framework/Compilation/Parser/Binding/Parser/ArrayInitializerExpression.cs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace DotVVM.Framework.Compilation.Parser.Binding.Parser
+{
+    [DebuggerDisplay("{DebuggerDisplay,nq}")]
+    public class ArrayInitializerExpression : BindingParserNode
+    {
+        public IList<BindingParserNode> ElementInitializers { get; private set; }
```
Comment (exyi) [inline src/Framework/Framework/Compilation/Parser/Binding/Parser/ArrayInitializerExpression.cs:10]:
```suggestion
        public List<BindingParserNode> ElementInitializers { get; private set; }
```

I'd prefer to use List or array. The IList interface is fairly broken - it contains the Add method, but arrays can be assigned to it, so it may fail randomly at runtime...

File: src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs [hunk 9]
```diff
@@ -47,6 +47,75 @@ public BindingParserNode ReadDirectiveValue()
             return first;
         }
 
+        public BindingParserNode ReadArrayInitializerValue()
+        {
+            var startIndex = CurrentIndex;
+            SkipWhiteSpace();
+
+            if (Peek() is BindingToken openBraceToken && openBraceToken.Type == BindingTokenType.OpenArrayBrace)
```
[the remark this answers is not in the document]
Comment (exyi) [reply src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs:55]: Actually, you can use `PeekType() == BindingTokenType.OpenArrayBrace`

File: src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs [hunk 10]
```diff
@@ -47,6 +47,75 @@ public BindingParserNode ReadDirectiveValue()
...
+                Read();
+                SkipWhiteSpace();
+                initializerExpressions.Add(ReadArrayInitializerValue());
+                SkipWhiteSpace();
+
+                while (Peek() is BindingToken comma && comma.Type == BindingTokenType.Comma)
+                {
+                    Read();
+                    SkipWhiteSpace();
+                    initializerExpressions.Add(ReadArrayInitializerValue());
```
Comment (exyi) [inline src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs:67]: I think you should also skip whitespace after the expression

File: src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs [hunk 11]
```diff
@@ -47,6 +47,75 @@ public BindingParserNode ReadDirectiveValue()
             return first;
         }
 
+        public BindingParserNode ReadArrayInitializerValue()
```
Comment (exyi) [inline src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs:50]: It seems that you gave this expression very low precedence (bellow OrElse), while it should be on par with other atomic expressions (constants, parenthesized exprs). For example `[1, 2, 3].Where(x => x % 2 == 0)` won't work this way (I think)

File: src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs (hunk 1 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs]: Added exceptions

File: src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs (hunk 2 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs:540]: I have tried `TypeConversions.ImplicitConversion` at first but it is intended for html attribute values and does such random magic as to be unhelpful. The numeric conversions are contraproductive here for example.

File: src/Framework/Framework/Compilation/ControlTree/IAbstractDirectiveAttributeReference.cs (hunk 5 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/ControlTree/IAbstractDirectiveAttributeReference.cs]: Added comments removed public

File: src/Framework/Framework/Compilation/ControlTree/Resolved/DirectiveCompilationService.cs (hunk 6 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/ControlTree/Resolved/DirectiveCompilationService.cs]: Suggestion accepted

File: src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+                wrapperType = CreateDymanicDeclaringType(wrapperType) ?? wrapperType;
+            }
+
+            return wrapperType;
+        }
+
+        protected virtual ITypeDescriptor? CreateDymanicDeclaringType(ITypeDescriptor? originalWrapperType)
+        {
+            var baseType = originalWrapperType?.CastTo<ResolvedTypeDescriptor>().Type ?? typeof(DotvvmMarkupControl);
+            AssemblyBuilder.DefineDynamicAssembly(new AssemblyName(Guid.NewGuid().ToString()), AssemblyBuilderAccess.Run);
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs:70]: Why?

File: src/Framework/Framework/Compilation/Directives/MarkupPageMetadata.cs [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        public IReadOnlyDictionary<string, IReadOnlyList<IAbstractDirective>> Directives { get; }
+        public ImmutableList<NamespaceImport> Imports { get; }
+        public IAbstractControlBuilderDescriptor? MasterPage { get; }
+        public ImmutableList<InjectedServiceExtensionParameter> InjectedServices { get; }
+        public ITypeDescriptor BaseType { get; }
+        public ITypeDescriptor? ViewModelType { get; }
+        public ViewModuleCompilationResult? ViewModuleResult { get; }
+        public ImmutableList<DotvvmProperty> Properties { get; }
+
+        public MarkupPageMetadata(
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/MarkupPageMetadata.cs:20]: How so?

File: src/Framework/Framework/Compilation/Directives/PropertyDeclarationDirectiveCompiler.cs [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+                result.Add((new ActualTypeReferenceBindingParserNode(attributeTypeReference), attributePropertyNameReference, initializer));
+            }
+            return result;
+        }
+
+        protected override ImmutableList<DotvvmProperty> CreateArtefact(IReadOnlyList<IAbstractPropertyDeclarationDirective> directives)
+        {
+            foreach (var directive in directives)
+            {
+                directive.DeclaringType = controlWrapperType;
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/PropertyDeclarationDirectiveCompiler.cs:96]: Directive compilation is a bit complicated now.

First we parse al the directives and resolve thehem according to the order in which they depend on each other. For example to create wrapper type we need to know what are the declared properties to be able to decide if we should create dynamic wrapper type. 

That means we have to parse and resolve `@property` directives before `@baseType` directive. So the order goes like this:
1. resolve `@property` directive, determine property name and type
2. using resolved properties from `@property` decide weather or not to create dynamic wraper type based on the base type  from @baseType directive.
3. Backfill the resolved wrapper type to the `@property` directives

TLDR: solution to a cyclic dependancy.

File: src/Framework/Framework/Compilation/ControlTree/Resolved/ResolvedPropertyDeclarationDirective.cs (hunk 7 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/ControlTree/Resolved/ResolvedPropertyDeclarationDirective.cs]: solved

File: src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+            var assemblyName = new AssemblyName(newDynamicAssemblyName);
+            var assemblyBuilder =
+                AssemblyBuilder.DefineDynamicAssembly(
+                    assemblyName,
+                    AssemblyBuilderAccess.Run);
+
+            // For a single-module assembly, the module name is usually
+            // the assembly name plus an extension.
+            var mb =
+                assemblyBuilder.DefineDynamicModule(newDynamicAssemblyName);
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs:82]: No idea how to do that.

File: src/Framework/Framework/Compilation/Directives/PropertyDeclarationDirectiveCompiler.cs [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+            }
+
+            return directives
+            .Where(d => d.PropertyType is ResolvedTypeDescriptor resolvedTypeDescriptor && resolvedTypeDescriptor.Type is not null)
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/PropertyDeclarationDirectiveCompiler.cs:100]: Neat, didnt know this one

File: src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+            if (fileName.EndsWith(".dotcontrol", StringComparison.Ordinal))
+            {
+                wrapperType = new ResolvedTypeDescriptor(typeof(DotvvmMarkupControl));
+            }
+            else
+            {
+                wrapperType = new ResolvedTypeDescriptor(typeof(DotvvmView));
+            }
+            return wrapperType;
+        }
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs:110]: Missed during refactoring, accepted

File: src/Framework/Framework/Compilation/Directives/MarkupDirectiveCompilerPipeline.cs [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+        public MarkupPageMetadata Compile(DothtmlRootNode dothtmlRoot, string fileName)
+        {
+            var directivesByName = dothtmlRoot.Directives.GroupBy(d => d.Name, StringComparer.OrdinalIgnoreCase).ToDictionary(d => d.Key, d => (IReadOnlyList<DothtmlDirectiveNode>)d.ToList());
+
+            var resolvedDirectives = new Dictionary<string, IReadOnlyList<IAbstractDirective>>();
+
+            var importCompiler = new ImportDirectiveCompiler(directivesByName, treeBuilder);
+            var importResult = importCompiler.Compile();
+            var imports = importResult.Artefact;
+            resolvedDirectives.AddIfAny(importCompiler.DirectiveName, importResult.Diractives);
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/MarkupDirectiveCompilerPipeline.cs:34]: fixed

File: src/Framework/Framework/Compilation/ControlTree/ControlTreeResolverBase.cs (hunk 4 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/ControlTree/ControlTreeResolverBase.cs:76]: 🤷‍♂️

File: src/Framework/Framework/Compilation/Parser/Binding/Parser/ArrayInitializerExpression.cs (hunk 8 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Parser/Binding/Parser/ArrayInitializerExpression.cs:10]: Other nodes use List so List it is, also the privet set has no reason, removed.

File: src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs (hunk 11 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs:50]: The question is: Should it work. I intentionally wanted to prevent "creative" uses of the array initializer ;)

File: src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs (hunk 10 above)
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs:67]: True, fixed

File: src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs]: It did nothing I see removed

File: src/Framework/Framework/Compilation/Directives/MarkupPageMetadata.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/MarkupPageMetadata.cs]: Ok

File: src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs (hunk 11 above)
Comment (exyi) [reply src/Framework/Framework/Compilation/Parser/Binding/Parser/BindingParser.cs:50]: Well, `([1, 2, 3]).Where(...)` will still work (AFAIK), so this isn't actually preventing anything

File: src/Framework/Framework/Compilation/Directives/PropertyDeclarationDirectiveCompiler.cs [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+            .Select(d => TryCreateDotvvmPropertyFromDirective(d))
+            .ToImmutableList();
+        }
+
+        protected virtual DotvvmProperty TryCreateDotvvmPropertyFromDirective(IAbstractPropertyDeclarationDirective propertyDeclarationDirective)
+        {
+            var propertyType = propertyDeclarationDirective.PropertyType.As<ResolvedTypeDescriptor>()!;
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/PropertyDeclarationDirectiveCompiler.cs]: fixed

File: src/Framework/Framework/Compilation/Directives/ViewModuleCompilationResult.cs [hunk 20]
```diff
@@ -0,0 +1,18 @@
...
+    public record ViewModuleCompilationResult
+    {
+        public JsExtensionParameter ExtensionParameter { get; }
+        public ViewModuleReferenceInfo Reference { get; }
+
+        public ViewModuleCompilationResult(JsExtensionParameter extensionParameter, ViewModuleReferenceInfo viewModuleReferenceInfo)
+        {
+            ExtensionParameter = extensionParameter;
+            Reference = viewModuleReferenceInfo;
+        }
```
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/ViewModuleCompilationResult.cs]: This metadata belongs solely to directive compilation, but shortened the record.

File: src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (Mylan719) [reply src/Framework/Framework/Compilation/Directives/BaseTypeDirectiveCompiler.cs]: Implemented

**Review by exyi — COMMENTED** [review]
Ok, I approve of this. However, I created the PR, so I can't give it a formal approval :sweat_smile:

**Review by Mylan719 — APPROVED** [review]
After review from @exyi and nd @acizmarik we concluded the PR is ready to merge
